### ISSUE TITLE
Get long-lived token dance in this lib

### ISSUE THREAD

**Issue by andresgutgon** [TiagoGrosso/instagram-graph-api-lib#67]
Hi, first of all thanks for this lib. Looks like can make my life easier : ). I have an idea on mind that requires to allow users to schedule when they want to publish content on their IG accounts. I see you explained [long-lived tokens in the readme](https://github.com/TiagoGrosso/instagram-graph-api-lib#access-token). 

I guess that is the kind of tokens I want to use. But reading through Facebook docs looks like the time these long-lived tokens are valid is also limited (like 60 days). And also can be expired by any reason at any time. 

What I'm think that could be done is a token dance like this:
```
1. User authorise my app with required permissions for publishing in their IG account
2. I get the User Access token and with this token I get the long-lived token
3. I can do schedule publishing at any time after user authorised my app
```

This would be the happy path. 

The non-happy paths are 2 on what I see:

### Long-lived token is about to expire
In this case always before using your lib I check expire date of the long-lived token I have stored. If is less than a day I rerequest server-to-server for a new long-lived token. And all is good 

### Long-lived token is invalidated or expired before expire date
For some reason the token I have stored is made invalid. In this case when I'm going to use your lib the request will throw an API error saying the token is not expired. In this case I read you can rerequest a new token with an expired token. In this case what I was thinking was to store on my side the publishing failed and let the user know about it. Do you think there is a better idea here?

### My question
Anyway. Those are my problems. Maybe I end up doing all of some of that. My question for you is if doing the `rerequest` dance for a long-lived token when using your lib would be interesting to be added to this code? Or do you think is better to do each one on our apps?

Not asking for you todo. Asking if you think it could be interesting and maybe I could contribute : )

Again, thanks for your work!

**Comment by TiagoGrosso** [comment]
Hey @andresgutgon , thanks for your interest!

First of all, let me just clarify that the "Long-Lived Token" expiration date depends on the type of token. If you generate a page token from a long-lived user token, it has no expiration date (check the last sentence on the section "Get a Page Access Token" on https://developers.facebook.com/docs/pages/access-tokens/).

Now, this token does not apply to every use case, of course, it only applies when you simply want to retrieve info about your own pages and not when you have an app that might access many users info, and that even seems to be your use case. So with this in mind, to answer your question: 

- I see this potentially happening to a lot of people and this library does aim to reduce the work of developers integrating with Instagram
- I haven't tested the "rerequest a new token with an expired token", but I'm assuming it works in a straightforward way
- I think most of the retry/refresh logic should be done on the app's side but the ability to request a new token could definitely be added here

So all in all, if you are up for it, I'm super happy to review the Pull Request. As you probably have seen, I work on this lib in batches, which is when I have a few days off from work, so I don't expect to have time to implement it myself any time soon, especially since I don't have an app ready to test this, I only have some page token for my own Instagram pages

**Comment by andresgutgon** [comment]
> First of all, let me just clarify that the "Long-Lived Token" expiration date depends on the type of token. If you generate a page token from a long-lived user token, it has no expiration date (check the last sentence on the section "Get a Page Access Token" on https://developers.facebook.com/docs/pages/access-tokens/).

Oh! I totally missed this sentence!

So if I do the oauth dance between me <-> the_user <-> Instagram(aka Facebook) I get a user access token.

In that moment I can do another dance 😂   to get the Instagram/Facebook PAGE_ID that the user wants to connect. Here I have a question. When the docs talk about "page". Are they referring to "Facebook page", "Instagram business account" or both? If they refer to the Instagram business account I guess I could to what is said in that paragraph in behalf of user with their "user-access_token" couldn't I? 

I mean [this part](https://developers.facebook.com/docs/pages/access-tokens/#get-a-page-access-token)

**Comment by TiagoGrosso** [comment]
If I recall correctly, the page token I'm referring to needs to be connected to a [Facebook App](https://developers.facebook.com/apps/).  Perhaps by using the [API Explorer](https://developers.facebook.com/tools/explorer/) this will be clearer. I think I'll also add this API Explorer to the docs

So if you are the administrator of a page (an Instagram business/creator page, connected to a Facebook page), you can create a Facebook App and you can use that to create a page token for that page, and that page token is the one that can be extended to have no expiration

This means that if you have an app where users have to log in, you'll never be able to generate a page token for them, only a user token. This is a bit confusing even for myself, as my use case for this lib was not to create an app that is open to other users 😅

**Comment by andresgutgon** [comment]
> This means that if you have an app where users have to log in, you'll never be able to generate a page token for them, only a user token

Ok yes, this is my use case. Is not only for myself. Is for anyone using the site and login with Facebook

**Comment by andresgutgon** [comment]
I think I found how to get long lived tokens (Almost forever tokens) for a page:

https://developers.facebook.com/docs/facebook-login/access-tokens/refreshing/#get-a-long-lived-user-access-token

It implies 2 calls:
1. First exchange a user-short-lived token by a user-long-lived-token (~60 days)
2. With that user-long-lived-token ask for user pages and get the page-long-lived-token

I think the strategy to make an app that can be used for any user would be to do this token dance and store that `page-long-lived-token` in the DB. This you can for example create an Instagram post on a cron job on behave of the user that authorised you to do it.

**Comment by TiagoGrosso** [comment]
Closed as #69 adds a request to get long lived tokens

**Comment by pburrows** [comment]
I want to add for anyone struggling with this: the only way in 2025 I was able to generate the Page Access Token with the required scopes was by using the API explorer linked above (https://developers.facebook.com/tools/explorer/)
